Local hardware buttons (needs testing and webui work)

miccgn opened the pull request:
As I am using a PCB that allows for three buttons to be connected to the ESP8266, I decided to play around with them.

I added support for:
- defining which pins to use for which button
- defining the level of the button for "released" and for "pressed" (some pins better are HIGH during boot of the ESP8266, so pressed=LOW).

Then I defined actions for button1 and button2:
Button1: toggle sleepmode
Button2: force the clock to be shown

I still have no idea what to do with button3. I could define two levels of brightness to be set in the options and jump between them?

I haven't made the settings visible in the options page yet. I suggest to create a new WebUI page labelled "Hardware", including all options for pin settings etc.

Work will be continued. What do you think?

hamster65 commented:
Nice! With this, I could connect a PIR sensor directly to the PixeIt. But it should be more universal and not trigger local things in the code. Instead, events should go to node-red / mqtt.
Having output pins to switch things on or off (for example the matrix power supply, or the mp3 module) would be another idea.

miccgn commented:
Output pins already exist. See the API function [setGpio](https://docs.bastelbunker.de/pixelit/api.html#set-gpio).

miccgn commented:
I would assign the DEFAULT values for the buttons pins to D0, D4, D5. 
As a result, pins would be used as follows by default:


| Pin | Default Usage |
|--- | ------------- |
| D0 | Btn1 |
| D1 | I2C/OneWire |
| D2 | Matrix |
| D3 | I2C |
| D4 | Btn2 |
| D5 | Btn3 |
| D6 |   |
| D7 | DFPlayer |
| D8 | DFPlayer |
| RX | Serial |
| TX | Serial |
| A0 | LDR |

miccgn commented:
> But it should be more universal and not trigger local things in the code. Instead, events should go to node-red / mqtt.

Idea: always make time of last btn press available via api and via mqtt.
Additionally, allow two actions being assigned via web interface:
- Toggle Sleep Mode
- Go to clock
-

miccgn commented:
I added option for customizing the action that should happen when a button is pressed. I also did some rework to the code to reorganize it.
Moreover, I added MQTT messages (whenever a button is pressed, the 'time' in millis() since boot will be published) and an API method returning that very time for all three buttons. By this, a client pulling states via API can differentiate between button presses (as the timestamp changes).
I also could use "real time" (timestamp in UTC) or just any random hash, or a counter counting the number of button presses. The API just needs anything that changes whenever a button is pressed.

Did not do anything to the WebUI yet.

miccgn commented:
I made a first draft for the web ui. But I have to admit: I can't test it, I can't run it and I have little knowledge of Vue.js.
@o0shojo0o here you need to jump in if possible - and if you like the whole buttons thing.
